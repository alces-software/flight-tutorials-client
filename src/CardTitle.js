// @flow
/*=============================================================================
 * Copyright (C) 2017 Stephen F. Norledge and Alces Software Ltd.
 *
 * This file is part of Flight Launch.
 *
 * All rights reserved, see LICENSE.txt.
 *===========================================================================*/
import React from 'react';
import classNames from 'classnames';

import CardTitleLogo from './CardTitleLogo';

const titleElemFromSize = (size) => {
  switch (size) {
    case 'medium':
      return 'h4';
    case 'large':
      return 'h2';
    case 'x-large':
      return 'h1';
    default:
      return 'h4';
  }
};

const subtitleElemFromSize = (size) => {
  switch (size) {
    case 'medium':
      return 'h6';
    case 'large':
      return 'h5';
    case 'x-large':
      return 'h4';
    default:
      return 'h6';
  }
};

type PropTypes = {
  icon? : React$Element<*>,
  logoUrl? : string,
  subtitle? : string | React$Element<*>,
  subtitleSize? : 'medium' | 'large' | 'x-large',
  title : string | React$Element<*>,
  titlePopoverText? : string,
  titleSize? : 'medium' | 'large' | 'x-large',
};

const defaultProps = {
  titleSize: 'x-large',
  subtitleSize: 'x-large',
};

const CardTitle = ({
  icon,
  logoUrl,
  subtitle,
  subtitleSize,
  title,
  titlePopoverText,
  titleSize,
}: PropTypes) => {
  const HT = titleElemFromSize(titleSize);
  const HST = subtitleElemFromSize(subtitleSize);
  const logoOrIcon = logoUrl ?
    <CardTitleLogo logoUrl={logoUrl} /> :
    icon;

  let titleAttribute;
  if (titlePopoverText != null) {
    titleAttribute = { title: titlePopoverText };
  } else {
    titleAttribute = typeof title === 'string' ? { title } : null;
  }

  return (
    <div className="card-title-block">
      <div className="card-title-block-inner">
        <div className="card-title-logo-container">
          <div className="card-title-logo-helper" />
          {logoOrIcon}
        </div>
        <HT className={classNames('card-title', titleSize)} {...titleAttribute}>
          {title}
        </HT>
      </div>
      <HST className={classNames('card-subtitle', subtitleSize)}>{subtitle}</HST>
    </div>
  );
};

CardTitle.defaultProps = defaultProps;

export default CardTitle;
