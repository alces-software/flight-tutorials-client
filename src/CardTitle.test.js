// @flow
/*=============================================================================
 * Copyright (C) 2017 Stephen F. Norledge and Alces Flight Ltd.
 *
 * This file is part of Flight Tutorials.
 *
 * All rights reserved, see LICENSE.txt.
 *===========================================================================*/
import React from 'react';
import { render } from 'react-dom';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import CardTitle from './CardTitle';

it('renders without crashing', () => {
  const node = document.createElement('div');
  render(<CardTitle title="My title" />, node);
});

const variants = {
  'default sized' : {
    title: 'My title',
    subtitle: 'My subtitle',
  },
  'large title and subtitle' : {
    title: 'My title',
    subtitle: 'My subtitle',
    titleSize: 'large',
    subtitleSize: 'large',
  },
  'x-large title and subtitle' : {
    title: 'My title',
    subtitle: 'My subtitle',
    titleSize: 'x-large',
    subtitleSize: 'x-large',
  },
  'medium title and subtitle' : {
    title: 'My title',
    subtitle: 'My subtitle',
    titleSize: 'medium',
    subtitleSize: 'medium',
  },
  'differing title and subtitle sizes' : {
    title: 'My title',
    subtitle: 'My subtitle',
    titleSize: 'x-large',
    subtitleSize: 'medium',
  },
  'with title popover text': {
    title: 'My title',
    subtitle: 'My subtitle',
    titlePopoverText: 'My title popover',
  },
  'with logo url': {
    title: 'My title',
    subtitle: 'My subtitle',
    logoUrl: 'http://localhost/my/logo.png',
  }
};

Object.keys(variants).forEach((name) => {
  it(`renders correctly: ${name}`, () => {
    const props = variants[name];
    const tree = renderer.create(
      <CardTitle {...props} />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe('title attribute', () => {
  it('sets a title attribute when title prop is a string', () => {
    const wrapper = shallow(
      <CardTitle title="My title" subtitle="Subtitle" />
    );

    expect(wrapper.find('.card-title')).toHaveProp('title', 'My title');
  });

  it('does not set a title attribute when title prop is a react node', () => {
    // This behaviour prevents the title being set to `[Object object]`.
    // Ideally, we'd extract the text from the title prop.
    const wrapper = shallow(
      <CardTitle title={<span>My title</span>} subtitle="Subtitle" />
    );

    expect(wrapper.find('.card-title')).not.toHaveProp('title', 'My title');
  });
});

it('does not crash when given an invalid titleSize', () => {
  const node = document.createElement('div');
  render(
    // Ignore flow error so that we can test that CardTitle will be robust
    // even if the tutorials contain bad data.
    // $FlowFixMe
    <CardTitle title="My title" subtitle="Subtitle" titleSize="not_valid" />,
    node
  );
});

it('does not crash when given an invalid subtitleSize', () => {
  const node = document.createElement('div');
  render(
    // Ignore flow error so that we can test that CardTitle will be robust
    // even if the tutorials contain bad data.
    // $FlowFixMe
    <CardTitle title="My title" subtitle="Subtitle" subtitleSize="not_valid" />,
    node
  );
});
