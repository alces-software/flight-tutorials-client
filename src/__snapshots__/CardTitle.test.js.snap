// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly: default sized 1`] = `
<div
  className="card-title-block"
>
  <div
    className="card-title-block-inner"
  >
    <div
      className="card-title-logo-container"
    >
      <div
        className="card-title-logo-helper"
      />
    </div>
    <h1
      className="card-title x-large"
      title="My title"
    >
      My title
    </h1>
  </div>
  <h4
    className="card-subtitle x-large"
  >
    My subtitle
  </h4>
</div>
`;

exports[`renders correctly: differing title and subtitle sizes 1`] = `
<div
  className="card-title-block"
>
  <div
    className="card-title-block-inner"
  >
    <div
      className="card-title-logo-container"
    >
      <div
        className="card-title-logo-helper"
      />
    </div>
    <h1
      className="card-title x-large"
      title="My title"
    >
      My title
    </h1>
  </div>
  <h6
    className="card-subtitle medium"
  >
    My subtitle
  </h6>
</div>
`;

exports[`renders correctly: large title and subtitle 1`] = `
<div
  className="card-title-block"
>
  <div
    className="card-title-block-inner"
  >
    <div
      className="card-title-logo-container"
    >
      <div
        className="card-title-logo-helper"
      />
    </div>
    <h2
      className="card-title large"
      title="My title"
    >
      My title
    </h2>
  </div>
  <h5
    className="card-subtitle large"
  >
    My subtitle
  </h5>
</div>
`;

exports[`renders correctly: medium title and subtitle 1`] = `
<div
  className="card-title-block"
>
  <div
    className="card-title-block-inner"
  >
    <div
      className="card-title-logo-container"
    >
      <div
        className="card-title-logo-helper"
      />
    </div>
    <h4
      className="card-title medium"
      title="My title"
    >
      My title
    </h4>
  </div>
  <h6
    className="card-subtitle medium"
  >
    My subtitle
  </h6>
</div>
`;

exports[`renders correctly: with logo url 1`] = `
<div
  className="card-title-block"
>
  <div
    className="card-title-block-inner"
  >
    <div
      className="card-title-logo-container"
    >
      <div
        className="card-title-logo-helper"
      />
      <img
        alt="Logo for card"
        className="card-title-logo"
        onLoad={[Function]}
        src="http://localhost/my/logo.png"
        style={
          Object {
            "display": "none",
          }
        }
      />
    </div>
    <h1
      className="card-title x-large"
      title="My title"
    >
      My title
    </h1>
  </div>
  <h4
    className="card-subtitle x-large"
  >
    My subtitle
  </h4>
</div>
`;

exports[`renders correctly: with title popover text 1`] = `
<div
  className="card-title-block"
>
  <div
    className="card-title-block-inner"
  >
    <div
      className="card-title-logo-container"
    >
      <div
        className="card-title-logo-helper"
      />
    </div>
    <h1
      className="card-title x-large"
      title="My title popover"
    >
      My title
    </h1>
  </div>
  <h4
    className="card-subtitle x-large"
  >
    My subtitle
  </h4>
</div>
`;

exports[`renders correctly: x-large title and subtitle 1`] = `
<div
  className="card-title-block"
>
  <div
    className="card-title-block-inner"
  >
    <div
      className="card-title-logo-container"
    >
      <div
        className="card-title-logo-helper"
      />
    </div>
    <h1
      className="card-title x-large"
      title="My title"
    >
      My title
    </h1>
  </div>
  <h4
    className="card-subtitle x-large"
  >
    My subtitle
  </h4>
</div>
`;
